File: packages/sdk-go/agentshield.go
package agentshield

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

type AgentShield struct {
	config    AgentShieldConfig
	evaluator *PolicyEvaluator
	client    *http.Client
}

func New(config AgentShieldConfig) *AgentShield {
	return &AgentShield{
		config:    config,
		evaluator: NewPolicyEvaluator(),
		client:    &http.Client{Timeout: 5 * time.Second},
	}
}

func (s *AgentShield) Guard(request ToolCallRequest, policy GuardrailPolicy) EvaluationResult {
	result := s.evaluator.Evaluate(request, policy)

	if !result.Allowed {
		if s.config.OnViolation != nil {
			s.config.OnViolation(result, request)
		}
		webhookURL := policy.WebhookURL
		if webhookURL == "" {
			webhookURL = s.config.WebhookURL
		}
		if webhookURL != "" {
			go s.dispatchWebhookAlert(result, request, webhookURL)
		}
	}

	if s.config.TelemetryURL != "" {
		go s.sendTelemetry(result, request)
	}

	return result
}

func (s *AgentShield) WrapTool(toolName string, toolFn func(map[string]interface{}) (interface{}, error), policy GuardrailPolicy) func(map[string]interface{}) (interface{}, error) {
	return func(params map[string]interface{}) (interface{}, error) {
		request := ToolCallRequest{
			ToolName: toolName,
			Params:   params,
		}
		evalResult := s.Guard(request, policy)
		if !evalResult.Allowed {
			return nil, fmt.Errorf("[AgentShield Blocked] %s", evalResult.Reason)
		}

		if policy.TimeoutMs > 0 {
			ctx, cancel := context.WithTimeout(context.Background(), time.Duration(policy.TimeoutMs)*time.Millisecond)
			defer cancel()

			type result struct {
				val interface{}
				err error
			}
			resultChan := make(chan result, 1)
			go func() {
				val, err := toolFn(params)
				resultChan <- result{val, err}
			}()

			select {
			case res := <-resultChan:
				return res.val, res.err
			case <-ctx.Done():
				return nil, fmt.Errorf("[AgentShield Timeout] Execution timed out after %dms", policy.TimeoutMs)
			}
		}

		return toolFn(params)
	}
}

func (s *AgentShield) dispatchWebhookAlert(result EvaluationResult, request ToolCallRequest, webhookURL string) {
	payload := map[string]interface{}{
		"event":      "AGENTSHIELD_VIOLATION_BLOCKED",
		"toolName":   request.ToolName,
		"reason":     result.Reason,
		"timestamp":  result.Timestamp,
		"params":     request.Params,
	}
	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", webhookURL, nil)
	req.Header.Set("Content-Type", "application/json")
	req.Body = http.NoBody
	req.GetBody = func() (io.ReadCloser, error) {
		return io.NopCloser(bytes.NewReader(body)), nil
	}
	s.client.Do(req)
}

func (s *AgentShield) sendTelemetry(result EvaluationResult, request ToolCallRequest) {
	payload := map[string]interface{}{
		"agentId":     request.AgentID,
		"toolName":    request.ToolName,
		"params":      request.Params,
		"actionTaken": result.ActionTaken,
		"reason":      result.Reason,
		"timestamp":   result.Timestamp,
	}
	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", s.config.TelemetryURL, nil)
	req.Header.Set("Content-Type", "application/json")
	req.Body = http.NoBody
	req.GetBody = func() (io.ReadCloser, error) {
		return io.NopCloser(bytes.NewReader(body)), nil
	}
	s.client.Do(req)
}

func Shield(toolName string, toolFn func(map[string]interface{}) (interface{}, error), policy GuardrailPolicy) func(map[string]interface{}) (interface{}, error) {
	globalShield := New(AgentShieldConfig{})
	return globalShield.WrapTool(toolName, toolFn, policy)
}

File: packages/sdk-rust/src/agentshield.rs
use reqwest::Client;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;
use std::time::Duration;
use tokio::time::timeout;

pub type BoxFuture<R> = Pin<Box<dyn Future<Output = R> + Send>>;

pub struct AgentShield {
    config: crate::types::AgentShieldConfig,
    evaluator: Arc<crate::evaluator::PolicyEvaluator>,
    http_client: Client,
}

impl AgentShield {
    pub fn new(config: crate::types::AgentShieldConfig) -> Self {
        let http_client = Client::builder()
            .timeout(Duration::from_secs(5))
            .build()
            .unwrap_or_default();

        Self {
            config,
            evaluator: Arc::new(crate::evaluator::PolicyEvaluator::new()),
            http_client,
        }
    }

    pub fn guard(&self, request: crate::types::ToolCallRequest, policy: crate::types::GuardrailPolicy) -> crate::types::EvaluationResult {
        let result = self.evaluator.evaluate(&request, &policy);

        if !result.allowed {
            if let Some(cb) = &self.config.on_violation {
                cb(&result, &request);
            }

            let webhook_url = policy.webhook_url.as_deref().or(self.config.webhook_url.as_deref());
            if let Some(url) = webhook_url {
                let result_clone = result.clone();
                let request_clone = request.clone();
                let url = url.to_string();
                let client = self.http_client.clone();
                tokio::spawn(async move {
                    let _ = Self::dispatch_webhook(client, url, result_clone, request_clone).await;
                });
            }
        }

        if let Some(telemetry_url) = &self.config.telemetry_url {
            let result_clone = result.clone();
            let request_clone = request.clone();
            let url = telemetry_url.clone();
            let client = self.http_client.clone();
            tokio::spawn(async move {
                let _ = Self::send_telemetry(client, url, result_clone, request_clone).await;
            });
        }

        result
    }

    pub fn wrap_tool<F, Fut, T, R>(&self, tool_name: &str, tool_fn: F, policy: crate::types::GuardrailPolicy) -> impl Fn(T) -> BoxFuture<Result<R, crate::error::AgentShieldError>> + Clone
    where
        F: Fn(T) -> Fut + Clone + Send + Sync + 'static,
        Fut: Future<Output = Result<R, crate::error::AgentShieldError>> + Send + 'static,
        T: Send + 'static + serde::Serialize,
        R: Send + 'static,
    {
        let shield = self.clone();
        let tool_name = tool_name.to_string();
        move |input| {
            let shield = shield.clone();
            let tool_name = tool_name.clone();
            let policy = policy.clone();
            let tool_fn = tool_fn.clone();
            Box::pin(async move {
                let policy_for_guard = policy.clone();
                let request = crate::types::ToolCallRequest {
                    tool_name: tool_name.clone(),
                    params: serde_json::to_value(&input).unwrap_or_else(|_| json!({ "input": input })).as_object().unwrap().clone().into_iter().collect(),
                    ..Default::default()
                };

                let eval_result = shield.guard(request, policy_for_guard);
                if !eval_result.allowed {
                    return Err(crate::error::AgentShieldError::Blocked(eval_result.reason.unwrap_or_default()));
                }

                let result_fut = tool_fn(input);
                
                if let Some(timeout_ms) = policy.timeout_ms {
                    match timeout(Duration::from_millis(timeout_ms), result_fut).await {
                        Ok(inner_result) => inner_result,
                        Err(_) => Err(crate::error::AgentShieldError::Timeout(timeout_ms)),
                    }
                } else {
                    result_fut.await
                }
            })
        }
    }

    async fn dispatch_webhook(client: Client, url: String, result: crate::types::EvaluationResult, request: crate::types::ToolCallRequest) -> Result<(), reqwest::Error> {
        let payload = json!({
            "event": "AGENTSHIELD_VIOLATION_BLOCKED",
            "toolName": request.tool_name,
            "reason": result.reason,
            "timestamp": result.timestamp,
            "params": request.params
        });
        client.post(&url).json(&payload).send().await?;
        Ok(())
    }

    async fn send_telemetry(client: Client, url: String, result: crate::types::EvaluationResult, request: crate::types::ToolCallRequest) -> Result<(), reqwest::Error> {
        let payload = json!({
            "agentId": request.agent_id,
            "toolName": request.tool_name,
            "params": request.params,
            "actionTaken": result.action_taken,
            "reason": result.reason,
            "timestamp": result.timestamp
        });
        client.post(&url).json(&payload).send().await?;
        Ok(())
    }
}

impl Clone for AgentShield {
    fn clone(&self) -> Self {
        Self {
            config: crate::types::AgentShieldConfig {
                api_key: self.config.api_key.clone(),
                environment: self.config.environment.clone(),
                telemetry_url: self.config.telemetry_url.clone(),
                webhook_url: self.config.webhook_url.clone(),
                on_violation: self.config.on_violation.clone(),
            },
            evaluator: self.evaluator.clone(),
            http_client: self.http_client.clone(),
        }
    }
}

impl Default for AgentShield {
    fn default() -> Self {
        Self::new(crate::types::AgentShieldConfig::default())
    }
}